### PR TITLE
sending raw stats to statshouse

### PR DESCRIPTION
**Pull request by troy4eg**
- [x]  worker stats (e.g. memory.rss_bytes, memory.malloc_mappep_bytes)
- [x] script stats (e.g. requests.outgoing_queries, memory.script_usage)

### CODE REVIEW

File: server/statshouse/add-metrics-batch.h [hunk 1]
```diff
@@ -28,3 +28,5 @@ struct StatsHouseAddMetricsBatch {
 };
 
 StatsHouseMetric make_statshouse_value_metric(std::string name, double value, const std::vector<std::pair<std::string, std::string>> &tags);
+
+StatsHouseMetric make_statshouse_value_metrics(std::string name, std::vector<double>&& value, const std::vector<std::pair<std::string, std::string>> &tags);
```
Comment (DrDet) [inline server/statshouse/add-metrics-batch.h:32]:
```suggestion
StatsHouseMetric make_statshouse_value_metrics(std::string &&name, std::vector<double> &&value, const std::vector<std::pair<std::string, std::string>> &tags);
```
Now it's copied always.

File: server/statshouse/add-metrics-batch.h [hunk 2]
```diff
@@ -28,3 +28,5 @@ struct StatsHouseAddMetricsBatch {
 };
 
 StatsHouseMetric make_statshouse_value_metric(std::string name, double value, const std::vector<std::pair<std::string, std::string>> &tags);
```
Comment (DrDet) [inline server/statshouse/add-metrics-batch.h:30]:
```suggestion
StatsHouseMetric make_statshouse_value_metric(std::string &&name, double value, const std::vector<std::pair<std::string, std::string>> &tags);
```

File: common/server/tl-stats-t.h [hunk 3]
```diff
@@ -33,4 +33,6 @@ class tl_stats_t : public stats_t {
     sb_printf(&sb, "%lld", value);
     sb_printf(&sb, "\n");
   }
+
+  void add_stat(const char *key [[maybe_unused]], std::vector<double> &&values [[maybe_unused]]) noexcept override {}
```
Comment (DrDet) [inline common/server/tl-stats-t.h:37]:
```suggestion
  void add_stat(const char *key [[maybe_unused]], std::vector<double> &&values [[maybe_unused]]) noexcept final { assert(false && "unimplemented"); }
```

File: common/stats/provider.h [hunk 4]
```diff
@@ -56,6 +57,16 @@ class stats_t {
     add_gauge_stat(stat_key, value);
   }
 
+  void add_gauge_stat(std::vector<double> &&values, const char *key1, const char *key2 = "") noexcept {
```
Comment (DrDet) [inline common/stats/provider.h:60]:
```suggestion
  void add_multiple_gauge_stats(std::vector<double> &&values, const char *key1, const char *key2 = "") noexcept {
```

File: server/statshouse/worker-stats-buffer.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    flush();
+  }
+  stats_buffer.add_stat(value);
+}
+
+void WorkerStatsBuffer::make_generic_metric(std::vector<StatsHouseMetric> &metrics, const char *name, GenericQueryStatKey stat_key, size_t worker_type,
+                                            const std::vector<tag> &tags) {
+  auto &stats_buffer = generic_query_stats[worker_type][static_cast<size_t>(stat_key)];
+  if (!stats_buffer.empty()) {
+    metrics.push_back(make_statshouse_value_metrics(name, stats_buffer.get_data_and_reset_buffer(), tags));
```
Comment (DrDet) [inline server/statshouse/worker-stats-buffer.cpp:64]: You forgot to update `last_send_time` here, let's update it in `flush` only

File: server/statshouse/worker-stats-buffer.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    std::vector<tag> tags;
+    tags.emplace_back("worker_type", i == static_cast<size_t>(WorkerType::general_worker) ? "general" : "job");
+    if (hostname != nullptr) {
+      tags.emplace_back("host", kdb_gethostname());
+    }
+
+    make_generic_metric(metrics, "kphp_requests_outgoing_queries", GenericQueryStatKey::outgoing_queries, i, tags);
+    make_generic_metric(metrics, "kphp_requests_outgoing_long_queries", GenericQueryStatKey::outgoing_long_queries, i, tags);
+    make_generic_metric(metrics, "kphp_requests_net_time", GenericQueryStatKey::script_time, i, tags);
+    make_generic_metric(metrics, "kphp_requests_script_time", GenericQueryStatKey::net_time, i, tags);
```
Comment (DrDet) [inline server/statshouse/worker-stats-buffer.cpp:90]: swap metric names, pls :)

File: server/statshouse/worker-stats-buffer.cpp [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  make_metric(metrics, "kphp_memory_job_common_request_usage", QueryStatKey::job_common_request_memory_usage, tags);
+  make_metric(metrics, "kphp_memory_job_common_request_real_usage", QueryStatKey::job_common_request_real_memory_usage, tags);
+
+  int offset = vk::tl::store_to_buffer(buffer, buffer_size, TL_STATSHOUSE_ADD_METRICS_BATCH);
+  offset = offset + vk::tl::store_to_buffer(buffer + offset, buffer_size - offset, static_cast<int>(vk::tl::statshouse::add_metrics_batch_fields_mask::ALL));
+  int len = vk::tl::store_to_buffer(buffer + offset, buffer_size - offset, metrics);
+  vk::singleton<StatsHouseClient>::get().send_metrics(buffer, offset + len);
+}
+
+void WorkerStatsBuffer::flush_if_needed() {
```
Comment (DrDet) [inline server/statshouse/worker-stats-buffer.cpp:116]: Why do we need to call this function in `cron`, if we always check time at `add_query_stat`?

**Review by DrDet — CHANGES_REQUESTED** [review]

File: server/statshouse/worker-stats-buffer.cpp (hunk 5 above)
Comment (troy4eg) [reply server/statshouse/worker-stats-buffer.cpp:67]: done

File: server/php-engine.cpp [hunk 8]
```diff
@@ -1665,6 +1667,7 @@ void init_all() {
 
   init_php_scripts();
   vk::singleton<ServerStats>::get().set_idle_worker_status();
+  vk::singleton<StatsHouseClient>::get();
```
[the remark this answers is not in the document]
Comment (troy4eg) [reply server/php-engine.cpp]: done

File: server/statshouse/statshouse-client.cpp [hunk 9]
```diff
@@ -104,25 +111,26 @@ bool StatsHouseClient::init_connection() {
 }
 
 void StatsHouseClient::send_metrics() {
```
[the remark this answers is not in the document]
Comment (troy4eg) [reply server/statshouse/statshouse-client.cpp]: done

**Review by DrDet — APPROVED** [review]

**Pull request merged**